admin/services/pin-generation.service.js

```diff
@@ -71,10 +69,10 @@ pinGenerationService.generatePupilPins = async (pupilsList) => {
   const data = Object.values(pupilsList || null)
   let pupils = []
   // fetch pupils
-  const ids = data.map(id => mongoose.Types.ObjectId(id))
+  const ids = data.map(id => id)
```

Comment:
this looks like a performance nightmare waiting to happen.
Can you not select all the pupils in one call at the end?